File: algorithmic-techniques/Recursion/Python/code.py
def factorial(n):
    if n == 0:
        return 1

    return n * factorial(n - 1)


def fibonacci(n):
    if n == 0:
        return 0

    if n == 1:
        return 1

    return fibonacci(n - 1) + fibonacci(n - 2)


def sum_recursive(n):
    if n == 0:
        return 0

    return n + sum_recursive(n - 1)


def array_sum(array, size):
    if size == 0:
        return 0

    return array[size - 1] + array_sum(array, size - 1)


def power(base, exponent):
    if exponent == 0:
        return 1

    return base * power(base, exponent - 1)


def gcd(a, b):
    if b == 0:
        return a

    return gcd(b, a % b)


def reverse_string_recursive(string, left, right):
    if left >= right:
        return

    string[left], string[right] = string[right], string[left]

    reverse_string_recursive(
        string,
        left + 1,
        right - 1
    )


def reverse_string(string):
    characters = list(string)

    reverse_string_recursive(
        characters,
        0,
        len(characters) - 1
    )

    return "".join(characters)


array = [2, 4, 6, 8]

print("Factorial:", factorial(5))

print("Fibonacci:", fibonacci(5))

print("Sum:", sum_recursive(5))

print("Array Sum:", array_sum(array, len(array)))

print("Power:", power(2, 5))

print("GCD:", gcd(48, 18))

print("Reversed String:", reverse_string("hello"))

File: algorithmic-techniques/Recursion/C/code.c
#include <stdio.h>
#include <string.h>

int factorial(int n)
{
    if (n == 0)
    {
        return 1;
    }

    return n * factorial(n - 1);
}

int fibonacci(int n)
{
    if (n == 0)
    {
        return 0;
    }

    if (n == 1)
    {
        return 1;
    }

    return fibonacci(n - 1) + fibonacci(n - 2);
}

int sum(int n)
{
    if (n == 0)
    {
        return 0;
    }

    return n + sum(n - 1);
}

int array_sum(const int *array, int size)
{
    if (size == 0)
    {
        return 0;
    }

    return array[size - 1] + array_sum(array, size - 1);
}

int power(int base, int exponent)
{
    if (exponent == 0)
    {
        return 1;
    }

    return base * power(base, exponent - 1);
}

int gcd(int a, int b)
{
    if (b == 0)
    {
        return a;
    }

    return gcd(b, a % b);
}

void reverse_string_recursive(
    char *string,
    int left,
    int right
)
{
    if (left >= right)
    {
        return;
    }

    char temp = string[left];
    string[left] = string[right];
    string[right] = temp;

    reverse_string_recursive(string, left + 1, right - 1);
}

void reverse_string(char *string)
{
    int length = strlen(string);

    reverse_string_recursive(
        string,
        0,
        length - 1
    );
}

int main(void)
{
    int array[] = {2, 4, 6, 8};
    int size = sizeof(array) / sizeof(array[0]);

    char string[] = "hello";

    printf("Factorial: %d\n", factorial(5));

    printf("Fibonacci: %d\n", fibonacci(5));

    printf("Sum: %d\n", sum(5));

    printf(
        "Array Sum: %d\n",
        array_sum(array, size)
    );

    printf("Power: %d\n", power(2, 5));

    printf("GCD: %d\n", gcd(48, 18));

    reverse_string(string);

    printf("Reversed String: %s\n", string);

    return 0;
}